### PR TITLE
Refactor g1 access

### PR DESCRIPTION
**Pull request by IntelOrca**
This makes `g1Elements` static so that all access to it goes via `gfx_get_g1_element` and `gfx_set_g1_element`. It also includes lots of `null` checks so that the game does not crash when there are missing g1 elements. This makes it a lot easier to get the game working without RCT2 assets.

Many palettes are currently stored in `g1.dat`, but we could later store them in the code so we do not need `g1.dat` for the game palette, shade and transparency tables.

### CODE REVIEW

File: src/openrct2/drawing/Sprite.cpp [hunk 1]
```diff
@@ -494,14 +504,14 @@ extern "C"
...
-     * image_id as below
+     * imageId (ebx)
+     * imageId as below
      * 0b_111X_XXXX_XXXX_XXXX_XXXX_XXXX_XXXX_XXXX image_type
      * 0b_XXX1_11XX_XXXX_XXXX_XXXX_XXXX_XXXX_XXXX image_sub_type (unknown pointer)
      * 0b_XXX1_1111_XXXX_XXXX_XXXX_XXXX_XXXX_XXXX secondary_colour
      * 0b_XXXX_XXXX_1111_1XXX_XXXX_XXXX_XXXX_XXXX primary_colour
      * 0b_XXXX_X111_1111_1XXX_XXXX_XXXX_XXXX_XXXX palette_ref
-     * 0b_XXXX_XXXX_XXXX_X111_1111_1111_1111_1111 image_id (offset to g1)
+     * 0b_XXXX_XXXX_XXXX_X111_1111_1111_1111_1111 imageId (offset to g1)
```
Comment (IntelOrca) [inline src/openrct2/drawing/Sprite.cpp]: Revert changes to doc-comment.

File: src/openrct2/drawing/Sprite.cpp [hunk 2]
```diff
@@ -522,7 +532,7 @@ extern "C"
 
     /*
     * rct: 0x0067A46E
-    * image_id (ebx) and also (0x00EDF81C)
+    * imageId (ebx) and also (0x00EDF81C)
```
Comment (IntelOrca) [inline src/openrct2/drawing/Sprite.cpp]: Revert changes to doc-comment.

**Comment by janisozaur** [comment]
Nice, we could probably tackle https://github.com/OpenRCT2/OpenRCT2/issues/3190 with that.

Load g1-less version, display some progress, reload the g1 after processing is done.

**Comment by IntelOrca** [comment]
@janisozaur I already have that working in a branch.

File: src/openrct2-ui/windows/Map.cpp [hunk 3]
```diff
@@ -880,29 +880,29 @@ static void window_map_paint(rct_window *w, rct_drawpixelinfo *dpi)
...
+    g1temp.width = MAP_WINDOW_MAP_SIZE;
+    g1temp.height = MAP_WINDOW_MAP_SIZE;
+    g1temp.x_offset = -8;
+    g1temp.y_offset = -8;
 
+    gfx_set_g1_element(SPR_TEMP, &g1temp);
     gfx_draw_sprite(dpi, SPR_TEMP, 0, 0, 0);
-
-    *g1_element = pushed_g1_element;
+    gfx_set_g1_element(SPR_TEMP, g1backup);
```
Comment (janisozaur) [inline src/openrct2-ui/windows/Map.cpp]: You have stored a `g1backup` as pointer and set `SPR_TEMP` to value of local stack `g1temp`. Won't it have invalidated the memory pointed at by `g1backup`?

Comment (janisozaur) [reply src/openrct2-ui/windows/Map.cpp]: I counted at least 3 times you do this

Comment (IntelOrca) [reply src/openrct2-ui/windows/Map.cpp]: This is just a table entry. The data pointer is different.

Comment (janisozaur) [reply src/openrct2-ui/windows/Map.cpp]: You are setting an item to itself. Either revert to previous behaviour or remove this call, as it is not doing anything.

Comment (IntelOrca) [reply src/openrct2-ui/windows/Map.cpp]: ok

File: src/openrct2/drawing/Sprite.cpp [hunk 4]
```diff
@@ -464,7 +466,15 @@ extern "C"
             }
 
             uint16 palette_offset = palette_to_g1_offset[palette_ref];
-            return g1Elements[palette_offset].offset;
+            auto g1 = gfx_get_g1_element(palette_offset);
+            if (g1 == NULL)
```
Comment (janisozaur) [inline src/openrct2/drawing/Sprite.cpp]: `nullptr`

File: src/openrct2-ui/windows/TrackList.cpp [hunk 5]
```diff
@@ -422,16 +422,17 @@ static void window_track_list_paint(rct_window *w, rct_drawpixelinfo *dpi)
...
+    rct_g1_element g1temp = { 0 };
+    g1temp.offset = _trackDesignPreviewPixels + (_currentTrackPieceDirection * TRACK_PREVIEW_IMAGE_SIZE);
+    g1temp.width = 370;
+    g1temp.height = 217;
+    g1temp.flags = G1_FLAG_BMP;
+
+    gfx_set_g1_element(SPR_TEMP, &g1temp);
     gfx_draw_sprite(dpi, 0, x, y, 0);
-    *substituteElement = tmpElement;
+    gfx_set_g1_element(SPR_TEMP, g1backup);
```
Comment (janisozaur) [inline src/openrct2-ui/windows/TrackList.cpp]: fix or remove

File: src/openrct2-ui/windows/Map.cpp [hunk 6]
```diff
@@ -880,29 +880,25 @@ static void window_map_paint(rct_window *w, rct_drawpixelinfo *dpi)
...
-    g1_element->x_offset = -8;
-    g1_element->y_offset = -8;
-    g1_element->flags = 0;
-
+    rct_g1_element g1temp = { 0 };
+    g1temp.offset = (uint8 *)_mapImageData;
+    g1temp.width = MAP_WINDOW_MAP_SIZE;
+    g1temp.height = MAP_WINDOW_MAP_SIZE;
+    g1temp.x_offset = -8;
+    g1temp.y_offset = -8;
```
Comment (janisozaur) [inline src/openrct2-ui/windows/Map.cpp]: misaligned assignments

File: src/openrct2-ui/windows/TrackDesignPlace.cpp [hunk 7]
```diff
@@ -467,16 +467,12 @@ static void window_track_place_paint(rct_window *w, rct_drawpixelinfo *dpi)
...
-        substituteElement->offset = _window_track_place_mini_preview;
-        substituteElement->width = TRACK_MINI_PREVIEW_WIDTH;
-        substituteElement->height = TRACK_MINI_PREVIEW_HEIGHT;
-        substituteElement->x_offset = 0;
-        substituteElement->y_offset = 0;
-        substituteElement->flags = 0;
+        rct_g1_element g1temp = { 0 };
+        g1temp.offset = _window_track_place_mini_preview;
+        g1temp.width = TRACK_MINI_PREVIEW_WIDTH;
+        g1temp.height = TRACK_MINI_PREVIEW_HEIGHT;
```
Comment (janisozaur) [inline src/openrct2-ui/windows/TrackDesignPlace.cpp]: misaligned assignments

File: src/openrct2-ui/windows/TrackList.cpp [hunk 8]
```diff
@@ -422,16 +422,17 @@ static void window_track_list_paint(rct_window *w, rct_drawpixelinfo *dpi)
...
-    substituteElement->x_offset = 0;
-    substituteElement->y_offset = 0;
-    substituteElement->flags = G1_FLAG_BMP;
+    auto g1backup = gfx_get_g1_element(SPR_TEMP);
+
+    rct_g1_element g1temp = { 0 };
+    g1temp.offset = _trackDesignPreviewPixels + (_currentTrackPieceDirection * TRACK_PREVIEW_IMAGE_SIZE);
+    g1temp.width = 370;
+    g1temp.height = 217;
+    g1temp.flags = G1_FLAG_BMP;
```
Comment (janisozaur) [inline src/openrct2-ui/windows/TrackList.cpp]: misaligned assignments

File: src/openrct2/drawing/Sprite.cpp [hunk 9]
```diff
@@ -746,39 +756,71 @@ extern "C"
...
             const uint32 idx = image_id - SPR_G2_BEGIN;
-            openrct2_assert(idx < _g2.header.num_entries,
-                            "Invalid entry in g2.dat requested, idx = %u. You may have to update your g2.dat.", idx);
+            if (idx >= _g2.header.num_entries)
+            {
+                // TODO I guess we still want to warn the user, but an assert is not very good.
+                //      Logging to stdout isn't ideal either as it would flood the console with
+                //      the same message.
+                // openrct2_assert(idx < _g2.header.num_entries,
+                //     "Invalid entry in g2.dat requested, idx = %u. You may have to update your g2.dat.", idx);
```
Comment (janisozaur) [inline src/openrct2/drawing/Sprite.cpp]: While I agree with your comment, this assert also helped me occasionally to know what the issue was.

I'd rather you left in the assert or replaced it with log. Yes, the latter might flood, but it is not in a valid state anyway and the only way to fix now-invisible parts is to fix the g2.

File: src/openrct2/drawing/Sprite.cpp [hunk 10]
```diff
@@ -746,39 +756,71 @@ extern "C"
...
-        rct_g1_element *g1_source = gfx_get_g1_element(image_id & 0X7FFFF);
+        const rct_g1_element * g1 = gfx_get_g1_element(image_id & 0X7FFFF);
         rct_size16 size = {};
-        if (g1_source != nullptr)
+        if (g1 != nullptr)
         {
-            size.width = g1_source->width;
-            size.height = g1_source->height;
+            size.width = g1->width;
+            size.height = g1->height;
```
Comment (janisozaur) [inline src/openrct2/drawing/Sprite.cpp]: misaligned assignments

**Comment by janisozaur** [comment]
You still violate your own code style guidelines, there are numerous misaligned assignments.

Other than that the code seems ok, but…

I tried testing this, but I'm a bit confused. You cannot really launch without `g1.dat`, as this gets checked at start and you haven't modified loading (or added a fallback) of g1.

I tried the next thing that came to my mind, which was to hardcode `gfx_get_g1_element` to return `nullptr`, but with this I either get completely black screen (I'm not using any title sequence) or when trying to load a park, I get a crash with this stack trace:
```
Thread 1 "openrct2" received signal SIGFPE, Arithmetic exception.
>>> bt full
#0  0x0000555555d9022a in sub_6A4101 (session=0x5555565eb4d0 <gPaintSession>, tile_element=0x555556751bf8 <gTileElements+623448>, height=128, ebp=6, word_F3F038=true, footpathEntry=0x55555cbbfe70, base_image_id=87176, imageFlags=0) at ../src/openrct2/paint/tile_element/path.c:411
        ride = 0x555556696890 <gRideList+17632>
        string_id = 1172
        string_width = 0
        scrollingMode = 1
        scroll = 0
        local_ebp = 6 '\006'
        direction = 1 '\001'
        boundBoxOffsets = {
          x = 32, 
          y = 29, 
          z = 130
        }
        imageId = 87282
        dword_F3EF80 = 128
#1  0x0000555555d8d017 in sub_6A3F61 (session=0x5555565eb4d0 <gPaintSession>, tile_element=0x555556751bf8 <gTileElements+623448>, bp=6, height=128, footpathEntry=0x55555cbbfe70, imageFlags=0, sceneryImageFlags=0, word_F3F038=true) at ../src/openrct2/paint/tile_element/path.c:631
        dpi = 0x7fffffffc978
        direction = 183 '\267'
        diagonal = 9 '\t'
        bl = 0 '\000'
#2  0x0000555555d8cbcf in path_paint_box_support (session=0x5555565eb4d0 <gPaintSession>, tileElement=0x555556751bf8 <gTileElements+623448>, height=128, footpathEntry=0x55555cbbfe70, hasFences=true, imageFlags=0, sceneryImageFlags=0) at ../src/openrct2/paint/tile_element/path.c:964
        edges = 6 '\006'
        boundBoxOffset = {
          x = 3, 
          y = 3
        }
        boundBoxSize = {
          x = 26, 
          y = 26
        }
        corners = 0 '\000'
        edi = 6
        imageId = 87233
        ax = 0
        supports = "\377\177\000"
#3  0x0000555555d8bb88 in path_paint (session=0x5555565eb4d0 <gPaintSession>, direction=1 '\001', height=128, tile_element=0x555556751bf8 <gTileElements+623448>) at ../src/openrct2/paint/tile_element/path.c:769
        word_F3F038 = true
        sceneryImageFlags = 0
        imageFlags = 0
        x = 1728
        y = 1760
        surface = 0x555556751cb0 <gTileElements+623632>
        bl = 16
        pathType = 7 '\a'
        footpathEntry = 0x55555cbbfe70
#4  0x00005555557b3558 in sub_68B3FB (session=0x5555565eb4d0 <gPaintSession>, x=1728, y=1760) at ../src/openrct2/paint/tile_element/tile_element.c:253
        direction = 1
        height = 128
        dword_9DE574 = {
          x = 1728, 
          y = 1760
        }
        segmentPositions = {[0] = {[0] = 0, [1] = 6, [2] = 2}, [1] = {[0] = 5, [1] = 4, [2] = 8}, [2] = {[0] = 1, [1] = 7, [2] = 3}}
        dpi = 0x7fffffffc978
        tile_element = 0x555556751bf8 <gTileElements+623448>
        rotation = 0 '\000'
        dx = 608
        bx = 1796
        element = 0x555556751be8 <gTileElements+623432>
        max_height = 1048
#5  0x00005555557b2e9c in tile_element_paint_setup (session=0x5555565eb4d0 <gPaintSession>, x=1728, y=1760) at ../src/openrct2/paint/tile_element/tile_element.c:61
No locals.
#6  0x00005555558c3b46 in paint_session_generate (session=0x5555565eb4d0 <gPaintSession>) at ../src/openrct2/paint/Paint.cpp:229
        dpi = 0x7fffffffc978
        mapTile = {
          x = 1728, 
          y = 1728
        }
        half_x = 0
        num_vertical_quadrants = 40
#7  0x000055555579da46 in viewport_paint_column (dpi=0x7fffffffc978, viewFlags=1024) at ../src/openrct2/interface/viewport.c:748
        session = 0x5555565eb4d0 <gPaintSession>
        ps = {
          image_id = 0, 
          {
            tertiary_colour = 0, 
            colour_image_id = 0
          }, 
          bound_box_x = 0, 
          bound_box_y = 0, 
          bound_box_z = 0, 
          bound_box_z_end = 0, 
          bound_box_x_end = 0, 
          bound_box_y_end = 0, 
          x = 0, 
          y = 0, 
          quadrant_index = 0, 
          flags = 0 '\000', 
          quadrant_flags = 0 '\000', 
          attached_ps = 0x0, 
          var_20 = 0x0, 
          next_quadrant_ps = 0x0, 
          sprite_type = 0 '\000', 
          var_29 = 0 '\000', 
          pad_2A = 0, 
          map_x = 0, 
          map_y = 0, 
          tileElement = 0x0
        }
#8  0x000055555579d990 in viewport_paint (viewport=0x55555664dae0 <g_viewport_list>, dpi=0x7fffffffca80, left=-1944, top=852, right=320, bottom=908) at ../src/openrct2/interface/viewport.c:731
        dpi2 = {
          bits = 0x5555575552cc "", 
          x = 0, 
          y = 852, 
          width = 32, 
          height = 56, 
          pitch = 1116, 
          zoom_level = 1
        }
        paintRight = 32
        viewFlags = 1024
        width = 2264
        height = 56
        bitmask = 65534
        x = 0
        y = 0
        dpi1 = {
          bits = 0x555557554f00 "", 
          x = -1944, 
          y = 852, 
          width = 2264, 
          height = 56, 
          pitch = 0, 
          zoom_level = 1
        }
        rightBorder = 320
#9  0x000055555579d63c in viewport_render (dpi=0x7fffffffca80, viewport=0x55555664dae0 <g_viewport_list>, left=-1944, top=853, right=320, bottom=909) at ../src/openrct2/interface/viewport.c:657
No locals.
#10 0x000055555573b1bb in window_main_paint (w=0x55555664e120 <g_window_list>, dpi=0x7fffffffca80) at ../src/openrct2-ui/windows/Main.cpp:99
No locals.
#11 0x00005555557ab5ed in window_event_paint_call (w=0x55555664e120 <g_window_list>, dpi=0x7fffffffca80) at ../src/openrct2/interface/window.c:2069
No locals.
#12 0x00005555557aa56e in window_draw_single (dpi=0x7fffffffca80, w=0x55555664e120 <g_window_list>, left=0, top=0, right=1132, bottom=28) at ../src/openrct2/interface/window.c:1718
        copy = {
          bits = 0x555557554f00 "", 
          x = 0, 
          y = 0, 
          width = 1132, 
          height = 28, 
          pitch = 0, 
          zoom_level = 0
        }
        overflow = 0
#13 0x00005555557aa042 in window_draw (dpi=0x7fffffffcb20, w=0x55555664e120 <g_window_list>, left=0, top=0, right=1132, bottom=28) at ../src/openrct2/interface/window.c:1614
        v = 0x55555664e120 <g_window_list>
#14 0x00005555557801b7 in window_draw_all (dpi=0x5555569a0988, left=0, top=0, right=1132, bottom=28) at ../src/openrct2/drawing/drawing.c:586
        w = 0x55555664e120 <g_window_list>
        windowDPI = {
          bits = 0x555557554f00 "", 
          x = 0, 
          y = 0, 
          width = 1132, 
          height = 28, 
          pitch = 0, 
          zoom_level = 0
        }
#15 0x000055555579f471 in viewport_shift_pixels (dpi=0x5555569a0988, w=0x55555664e600 <g_window_list+1248>, viewport=0x55555664dae0 <g_viewport_list>, x_diff=689, y_diff=1433) at ../src/openrct2/interface/viewport.c:402
        orignal_w = 0x55555664e120 <g_window_list>
        left = 0
        right = 1132
        top = 0
        bottom = 28
#16 0x000055555579d3c0 in viewport_move (x=-1944, y=853, w=0x55555664e120 <g_window_list>, viewport=0x55555664dae0 <g_viewport_list>) at ../src/openrct2/interface/viewport.c:482
        dpi = 0x5555569a0988
        zoom = 2 '\002'
        x_diff = 689
        y_diff = 1433
        view_copy = {
          width = 1132, 
          height = 752, 
          x = 0, 
          y = 0, 
          view_x = -1944, 
          view_y = 853, 
          view_width = 2264, 
          view_height = 1504, 
          flags = 1024, 
          zoom = 1 '\001', 
          var_11 = 0 '\000', 
          visibility = 0 '\000'
        }
        eax = 0
#17 0x000055555579ccf6 in viewport_update_position (window=0x55555664e120 <g_window_list>) at ../src/openrct2/interface/viewport.c:600
        viewport = 0x55555664dae0 <g_viewport_list>
        x = -1944
        y = 853
        z = 128
        at_map_edge_x = 0
        at_map_edge_y = 0
#18 0x000055555578f464 in game_load_init () at ../src/openrct2/game.c:1241
        zoomDifference = 1 '\001'
        mainWindow = 0x55555664e120 <g_window_list>
        intent = 0x7fffffffcce0
#19 0x000055555580e789 in OpenRCT2::Context::LoadParkFromStream (this=0x555556ca4550, stream=0x7fffffffce28, path="/home/janisozaur/Downloads/dome-roof-on_zoom1.sv6", loadTitleScreenFirstOnFail=true) at ../src/openrct2/Context.cpp:415
        parkImporter = std::unique_ptr<IParkImporter> containing 0x7fffbebde010
        result = {
          Error = PARK_LOAD_ERROR_OK, 
          MissingObjects = std::vector of length 0, capacity 0, 
          Flag = 0 '\000'
        }
        info = {
          Type = FILE_TYPE::SAVED_GAME, 
          Version = 120001
        }
#20 0x000055555580e51f in OpenRCT2::Context::LoadParkFromFile (this=0x555556ca4550, path="/home/janisozaur/Downloads/dome-roof-on_zoom1.sv6", loadTitleScreenOnFail=true) at ../src/openrct2/Context.cpp:378
        fs = {
          <IStream> = {
            _vptr$IStream = 0x555556211808 <vtable for FileStream+16>
          }, 
          members of FileStream: 
          _file = 0x555557720ef0, 
          _ownsFilePtr = true, 
          _canRead = true, 
          _canWrite = false, 
          _disposed = false, 
          _fileSize = 1031211
        }
#21 0x0000555555810196 in OpenRCT2::Context::Launch (this=0x555556ca4550) at ../src/openrct2/Context.cpp:546
        ex = <error reading variable>
#22 0x000055555580dea9 in OpenRCT2::Context::RunOpenRCT2 (this=0x555556ca4550, argc=3, argv=0x7fffffffd118) at ../src/openrct2/Context.cpp:158
No locals.
#23 0x00005555556ed94c in main (argc=3, argv=0x7fffffffd118) at ../src/openrct2-ui/Ui.cpp:60
        env = 0x5555568f7530
        audioContext = 0x555556907b90
        uiContext = 0x55555693cd80
        context = 0x555556ca4550
        runGame = 1
```

Either a bug/unhandled case or I did not understand fully how am I supposed to make use of this?

**Comment by IntelOrca** [comment]
It might make more sense if you try this branch:
https://github.com/IntelOrca/OpenRCT2/tree/spike/loading-screen

**Comment by IntelOrca** [comment]
I only ever aligned assignments outside routines. I never did it inside code and I don't like it - it is just too fiddly when coding.

**Pull request merged**